presto-geospatial/src/main/java/com/facebook/presto/geospatial/BingTileFunctions.java

```diff
@@ -0,0 +1,312 @@
+
+import static com.facebook.presto.geospatial.GeometryType.GEOMETRY_TYPE_NAME;
+import static com.facebook.presto.geospatial.GeometryUtils.POINT;
+import static com.facebook.presto.geospatial.GeometryUtils.deserialize;
+import static com.facebook.presto.geospatial.GeometryUtils.serialize;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static java.lang.String.format;
+
+/**
+ * A set of functions to convert between geometries and Bing tiles.
+ *
+ * See https://msdn.microsoft.com/en-us/library/bb259689.aspx for the description of the Bing tiles.
+ */
+public class BingTileFunctions
+{
+    private static final int TILE_PIXELS = 256;
+    private static final double MAX_LATITUDE = 85.05112878;
+    private static final double MIN_LATITUDE = -85.05112878;
+    private static final double MIN_LONGITUDE = -180;
+    private static final double MAX_LONGITUDE = 180;
+    private static final int MAX_ZOOM_LEVEL = 23;
+
+    private BingTileFunctions() {}
+
+    @Description("Creates a Bing tile from XY coordinates and zoom level")
+    @ScalarFunction("bing_tile")
+    @SqlType(BingTileType.NAME)
+    public static long toBingTile(@SqlType(StandardTypes.BIGINT) long tileX, @SqlType(StandardTypes.BIGINT) long tileY, @SqlType(StandardTypes.BIGINT) long zoomLevel)
+    {
+        checkZoomLevel(zoomLevel);
+        checkXYCoordinate(tileX, zoomLevel);
+        checkXYCoordinate(tileY, zoomLevel);
+
+        return new BingTile((int) tileX, (int) tileY, (int) zoomLevel).encode();
+    }
+
+    @Description("Given a Bing tile returns its Quad key")
+    @ScalarFunction("bing_tile_quadkey")
+    @SqlType(StandardTypes.VARCHAR)
+    public static Slice toQuadKey(@SqlType(BingTileType.NAME) long input)
+    {
+        return Slices.utf8Slice(BingTile.decode(input).toQuadKey());
+    }
+
+    @Description("Given a Bing tile, returns XY coordinates of the tile")
+    @ScalarFunction("bing_tile_coordinates")
+    @SqlType("array(bigint)")
+    public static Block bingTileCoordinates(@SqlType(BingTileType.NAME) long input)
+    {
+        BingTile tile = BingTile.decode(input);
+
+        BlockBuilder tileBlockBuilder = BIGINT.createBlockBuilder(new BlockBuilderStatus(), 2);
+        BIGINT.writeLong(tileBlockBuilder, tile.getX());
+        BIGINT.writeLong(tileBlockBuilder, tile.getY());
+
+        return tileBlockBuilder.build();
+    }
+
+    @Description("Given a Bing tile, returns zoom level of the tile")
+    @ScalarFunction("bing_tile_zoom_level")
+    @SqlType(StandardTypes.BIGINT)
+    public static long bingTileZoomLevel(@SqlType(BingTileType.NAME) long input)
+    {
+        return BingTile.decode(input).getZoomLevel();
+    }
+
+    @Description("Creates a Bing tile from a quad key")
+    @ScalarFunction("bing_tile")
+    @SqlType(BingTileType.NAME)
+    public static long toBingTile(@SqlType(StandardTypes.VARCHAR) Slice quadKey)
```

Comment:
I thought we were going to add a type for QuadKey?

Reply:
@electrum , I might have misunderstood you. In the earlier version I used quadkey string to represent a tile. In this PR, I introduced BingTileType serialized as long to represent a tile during query execution. The tiles still need to be persisted as quadkey strings or (X, Y, zoomLevel) tuples in Hive. I experimented with a cast from tile to varchar, but found the syntax too verbose and overall confusing. 

cast(point_to_bing_tile(ST_Point(23, 34.0), 15) as varchar) vs. bing_tile_quadkey(ST_Point(23, 34.0), 15))